dvc/tree/ssh/__init__.py

```diff
@@ -174,7 +174,11 @@ def isfile(self, path_info):
     def walk_files(self, path_info, **kwargs):
         with self.ssh(path_info) as ssh:
             for fname in ssh.walk_files(path_info.path):
-                yield path_info.replace(path=fname)
+                try:
+                    yield path_info.replace(path=fname)
+                except GeneratorExit:
+                    # raised when generator is dropped in the middle of walk
+                    break
```

Comment:
Great catch!

I guess the reason is https://github.com/iterative/dvc/blob/de01b11483c11d1b0cf55caf264b7bc65f3c8347/dvc/tree/pool.py#L14 , and not this particular piece of code. Also, this bug might affect other remote types as well, so even more reason to fix it there.

Reply:
The line that you pointed is responsible for dropping connections, that was used when some unhandled exception was raised. Maybe it's a bit aggressive but imo it's not a bad solution - possibly some bad state of connection was a cause for this exception?
The code that I fixed was buggy for sure. It's completely valid to start iterating `walk_files()` generator and abandon it without iterating through all elements. But without my fix, such a usage raises an exception, which is wrong.

Reply:
It should be valid to not catch `GeneratorExit` in a generator though, since `GeneratorExit` isn't even an error condition (which is why it inherits from `BaseException` and not `Exception`). Normally you catch `GeneratorExit` inside a generator function if you have to run cleanup code before exiting the function. We are not doing at all here, so this empty `except GeneratorExit` looks strange.

Basically, a generator raising `GeneratorExit` on `break` is expected behavior. The bug is that we incorrectly treat this expected behavior as an error in the connection pool wrapper.

The fix for this should probably be to catch `Exception`, and maybe also `KeyboardInterrupt`, instead of `BaseException` in the connection pool code.

Reply:
Ok, you both are right. I misunderstood a bit how `GeneratorExit` works in this case.

Reply:
Maybe @Suor could shed some light on the reasoning behind BaseException in get_connection?

Reply:
Not sure about here, but on pool we should watch for everything, including. The logic is clear:
- we pop a connection from a pool
- if something happened during that connection use then we close it
- if connection was used successfully we return it to the pool
- we always reraise so that behavior is not altered